hugo/content/docs/ast-types.md

```diff
@@ -3,4 +3,338 @@ title: "AST Type Inference"
+```
+This result in the creation of a single interface in the AST 'merging' the two parser rules with non-common properties made optional:
+```
+interface MyType extends AstNode {
+    count?: number
+    name: string
+}
+```
+
+### Terminal Rules
+Terminal rules are linked to built-in types in the AST. They do not result in AST types on their own but determine the type of properties in AST types inferred from parser rule:
+```
+terminal INT returns number: /[0-9]+/;
+terminal ID returns string: /[a-zA-Z_][a-zA-Z0-9_]*/;
+
+X: name=ID count=INT;
+
+// generates:
+interface X extends AstNode {
+    name: string
+    count: number
+}
+```
+The property `name` is of type `string` because the terminal rule `ID` is linked to the built-in type `string`, and the property `count` is of type `number` because the terminal rule `INT` is linked to the built-in type `number`.
+
+### Data type rules
+Data type rules are similar to terminal rules in the sense that they determine the type of properties in AST types inferred from a parser rules. However, they lead to the creation of type aliases for built-in types in the AST:
+```
+QualifiedName returns string: ID '.' ID;
+
+X: name=QualifiedName;
+
+// generates:
+type QualifiedName = string;
+
+interface X extends AstNode {
+    name: string
+}
+```
+
+### Assignments
+There are three available kinds of [assignments](../grammar-language/#assignments) in a parser rule:
+
+1. `=` for assigning a **single value** to a property, resulting in the property's type to be derived from the right side of the assignment.
+2. `+=` for assigning **multiple values** to a property, resulting in the property's type to be an array of the right side of the assignment.
+3. `?=` for assigning a **boolean** to a property, resulting in the property's type to be a `boolean`.
+
+```
+X: name=ID numbers+=INT (numbers+=INT)* isValid?='valid'?;
+
+// generates:
+interface X extends AstNode {
+    name: string
+    numbers: Array<number>
+    isValid: boolean
+}
+```
+
+The right side of an assignment can be any of the following:
+* A terminal rule or a data type rule, which results in the type of the property to be a built-in type.
+* A parser rule, which results is the type of the property to be the type of the parser rule.
+* A cross-reference, which results in the type of the property to be a *Reference* to the type of the cross-reference.
+* An alternative of all the above, which results in the type of the property to be a type union of all the types in the alternative.
+
+```
+X: 'x' name=ID;
+
+Y: crossValue=[X:ID] alt=(INT | X | [X:ID]);
+
+// generates:
+interface X extends AstNode {
+    name: string
+}
+
+interface Y extends AstNode {
+    crossValue: Reference<X>
+    alt: number | X | Reference<X>
+}
+```
+
+### Unassigned Rule Calls
+A parser rule does not necessarily need to have assignment, but can also only contain *unassigned rule calls*. These kind of rules can be used to change the types' hierarchy.
```

Comment:
A parser rule does not necessarily need to have **assignments**. **It may** also **contain only** *unassigned rule calls*.